torchsde/_core/base_sde.py

```diff
@@ -66,21 +66,21 @@ def __init__(self, sde, fast_dg_ga_jvp_column_sum=False):
     ########################################
 
     def g_prod_diagonal(self, t, y, v):
-        return self._base_sde.g(t, y) * v
+        return self.g(t, y) * v
 
     def g_prod_default(self, t, y, v):
-        return misc.batch_mvp(self._base_sde.g(t, y), v)
+        return misc.batch_mvp(self.g(t, y), v)
 
     ########################################
     #               gdg_prod               #
     ########################################
 
     # Computes: sum_{j, l} g_{j, l} d g_{j, l} d x_i v_l.
     def gdg_prod_default(self, t, y, v):
-        requires_grad = torch.is_grad_enabled()
+        requires_grad = torch.is_grad_enabled() and (t.requires_grad or y.requires_grad or v.requires_grad)
```

Comment:
I'm a little worried about this line. Consider the scenario where we set `y0` to not require gradients. By current standards, neither `t` nor `v` should require gradients. So in the backprop-through-solver setting, `requires_grad` evaluates to False. 

The problematic scenario is then solving an SDE with parameters that require gradients. Since `requires_grad` is False, we don't create the extra graph backpropping through `vg_dg_vjp` towards the parameters. However, in the ideal situation, we should still backprop through this node so that our models get better gradients. 

Overall, I think this possible failure case indicates that we should have a numerical test for backprop-through-solver comparing against finite differences.

Reply:
I think it's worth noting that we currently don't (a) backprop-through-the-solver with the adjoint, nor do we (b) differentiate wrt parameters without also differentiating wrt y, so as the intersection of these two null events we at least shouldn't be any running into any issues with the current codebase.

That said, this is a subtle point that we'd probably never catch in the future if this scenario does come up, so I agree that I'd prefer to do something about it now.

I'm not completely sure about a fix.
The current `requires_grad` is meant to correspond to "if I hadn't forced `y` to have a gradient, would `some_func(t, y, v)` require gradient?"
I'm -1 on any approach that just fixes it to `True`, as else we'll be leaking graphs from the function.
I think my proposed fix would be to:
- If `y` requires gradient, set `requires_grad` to True. (As before.)
- If `y` doesn't require gradient, then we're going to be forcing it to require gradient. Then we call `some_func(t, y, v)` as before. Prior to calling `vjp` we manually traverse its autograd history (via `.grad_fn`, `.next_functions`) and set `requires_grad = not (is y the only leaf)`.

Not a huge fan of that but I don't have any better ideas. Thoughts?

Reply:
> I think it's worth noting that we currently don't (a) backprop-through-the-solver with the adjoint, nor do we (b) differentiate wrt parameters without also differentiating wrt y, so we at least shouldn't be any running into any issues with the current codebase.

I don't think the line I'm specifically referring to has much to do with adjoints. `gdg_prod` is part of `ForwardSDE`, so in the backprop through solver case, it would get called when we choose, say, the Milstein method. 

More specifically, I think this will cause a problem in the first step of sdeint (w/o adjoints), where `y0` doesn't require gradient. The later steps shouldn't be affected by too much, since the typical `y` beyond `y0` should require gradients due to being sums of tensors that require gradient with tensors that don't.

Reply:
Oh sorry, you're commenting on base_sde rather than adjoint_sde. You're right, this would break Milstein.

I think the analysis for the forward case is the same as the backward case that I actually discussed, though, so I think I'd end up proposing the same fix. I'll see what you say.

Unrelatedly, I've updated `test_adjoint.py` to pytest. Scalar noise+Ito SDE seems to be failing the tests btw. (It wasn't tested previously.) I've not dug into why that is - I know this isn't a supported case on master but I thought it was in now?

Reply:
I think the issue I mentioned is mostly about when `sde.parameters()` is non-empty. So I would perhaps add a literal that checks whether `ForwardSDE` has any parameters that require gradients. 

I'll fix the Scalar noise + Ito case.

Reply:
So that relies on the implicit assumption that everything going into the function is a parameter. That needn't be the case, c.f. our discussion a bit back on contextualisation.